### PR TITLE
Enhance extrapolation module

### PR DESCRIPTION
**Pull request by KrzysztofB-1qbit**
- Fixed bug in richardson_analytical (added a sign change depending on number of points, agrees with the numerical extrapolation)
- Added optional error estimation (standard quadratic error propagation)
- Added relevant tests
- Changed order of arguments to keep energies and errors next to each other

### CODE REVIEW

File: tangelo/toolboxes/post_processing/extrapolation.py [hunk 1]
```diff
@@ -71,41 +77,53 @@ def extrapolation(energies, coeffs, taylor_order=None):
...
+        float: Error estimation for extrapolated energy
     """
     Eh = np.array(energies)
     ck = np.array(coeffs)
-    x = np.array([np.prod(ai/(a - ai)) for i, a in enumerate(ck)
-                      for ai in [np.delete(ck, i)]])
-    return np.dot(Eh, x)
+    n = len(ck)
+    x = np.array([(-1)**(n-1) * np.prod(ai/(a - ai)) for i, a in enumerate(ck) for ai in [np.delete(ck, i)]])
+
```
Comment (JamesB-1qbit) [inline tangelo/toolboxes/post_processing/extrapolation.py:118]: I think you can also write this as.
```python3
x = np.array([np.prod(ai/(ai - a)) for i, a in enumerate(ck) for ai in [np.delete(ck, i)]])
```

**Review by JamesB-1qbit — APPROVED** [review]
One suggestion.

File: tangelo/toolboxes/post_processing/extrapolation.py (hunk 1 above)
Comment (KrzysztofB-1qbit) [reply tangelo/toolboxes/post_processing/extrapolation.py:118]: That does seem like a simpler solution. Weird, I thought I tried that at some point. Changed

**Review by ValentinS4t1qbit — CHANGES_REQUESTED** [review]
Minor changes.

I see some redundant code in these functions: if you foresee that refactoring is useful to ensure these functions stay consistent even if we change them in the future, please consider encapsulating this code in a function with a `_` before the name. This is not a requirement, you make the decision.

File: tangelo/toolboxes/post_processing/extrapolation.py [hunk 2]
```diff
@@ -16,53 +16,59 @@
...
+def richardson(coeffs, energies, stderr=None, estimate_exp=False):
     """
     General, DIIS-like extrapolation procedure as found in
     Nature 567, 491-495 (2019) [arXiv:1805.04492]
 
     Args:
-        energies (array-like): Energy expectation values for amplified noise rates
         coeffs (array-like): Noise rate amplification factors
+        energies (array-like): Energy expectation values for amplified noise rates
+        stderr (array-like): Energy standard error estimates
```
[the remark this answers is not in the document]
Comment (KrzysztofB-1qbit) [reply tangelo/toolboxes/post_processing/extrapolation.py:44]: done

File: tangelo/toolboxes/post_processing/extrapolation.py [hunk 3]
```diff
@@ -16,53 +16,59 @@
...
 
     Returns:
         float: Extrapolated energy
+        float: Error estimation for extrapolated energy
     """
     if estimate_exp is False:
```
[the remark this answers is not in the document]
Comment (KrzysztofB-1qbit) [reply tangelo/toolboxes/post_processing/extrapolation.py:50]: done

File: tangelo/toolboxes/post_processing/extrapolation.py [hunk 4]
```diff
@@ -16,53 +16,59 @@
...
+def extrapolation(coeffs, energies, stderr=None, taylor_order=None):
     """
     General, DIIS-like extrapolation procedure as found in
     Nature 567, 491-495 (2019) [arXiv:1805.04492]
 
     Args:
-        energies (array-like): Energy expectation values for amplified noise rates
         coeffs (array-like): Noise rate amplification factors
+        energies (array-like): Energy expectation values for amplified noise rates
+        stderr (array-like): Energy standard error estimates
```
[the remark this answers is not in the document]
Comment (KrzysztofB-1qbit) [reply tangelo/toolboxes/post_processing/extrapolation.py:66]: done

File: tangelo/toolboxes/post_processing/extrapolation.py [hunk 5]
```diff
@@ -71,55 +77,70 @@ def extrapolation(energies, coeffs, taylor_order=None):
...
 
-def richardson_with_exp_estimation(energies, coeffs):
+def richardson_with_exp_estimation(coeffs, energies, stderr=None):
     """
     Richardson extrapolation by recurrence, with exponent estimation
 
     Args:
         energies (array-like): Energy expectation values for amplified noise rates
         coeffs (array-like): Noise rate amplification factors
+        stderr (array-like): Energy standard error estimates
```
[the remark this answers is not in the document]
Comment (KrzysztofB-1qbit) [reply tangelo/toolboxes/post_processing/extrapolation.py:132]: done

**Review by ValentinS4t1qbit — APPROVED** [review]
Thank you Krzys !

**Pull request merged**